File: worksteal.go
package ual

import (
	"sync"
	"sync/atomic"
)

// ============================================================================
// Traditional Work-Stealing (Chase-Lev style deque)
// As taught in computer science - lock-free owner, locked steal
// ============================================================================

// Task represents a unit of work
type Task struct {
	ID   int64
	Data []byte
}

// WSDeque is a work-stealing deque (traditional implementation)
// Owner pushes/pops from bottom (LIFO), thieves steal from top (FIFO)
type WSDeque struct {
	tasks  []Task
	bottom int64 // atomic - owner's end
	top    int64 // atomic - thief's end
	mu     sync.Mutex // for steal synchronization
}

// NewWSDeque creates a traditional work-stealing deque
func NewWSDeque(capacity int) *WSDeque {
	return &WSDeque{
		tasks:  make([]Task, capacity),
		bottom: 0,
		top:    0,
	}
}

// Push adds a task (owner only, LIFO end)
func (d *WSDeque) Push(t Task) bool {
	b := atomic.LoadInt64(&d.bottom)
	top := atomic.LoadInt64(&d.top)
	
	if b-top >= int64(len(d.tasks)) {
		return false // full
	}
	
	d.tasks[b%int64(len(d.tasks))] = t
	atomic.StoreInt64(&d.bottom, b+1)
	return true
}

// Pop removes a task (owner only, LIFO end)
func (d *WSDeque) Pop() (Task, bool) {
	b := atomic.LoadInt64(&d.bottom) - 1
	atomic.StoreInt64(&d.bottom, b)
	
	top := atomic.LoadInt64(&d.top)
	
	if top <= b {
		// Non-empty
		t := d.tasks[b%int64(len(d.tasks))]
		if top == b {
			// Last element - race with steal
			if !atomic.CompareAndSwapInt64(&d.top, top, top+1) {
				// Lost race
				atomic.StoreInt64(&d.bottom, b+1)
				return Task{}, false
			}
			atomic.StoreInt64(&d.bottom, b+1)
		}
		return t, true
	}
	
	// Empty
	atomic.StoreInt64(&d.bottom, top)
	return Task{}, false
}

// Steal takes a task (thief, FIFO end)
func (d *WSDeque) Steal() (Task, bool) {
	d.mu.Lock()
	defer d.mu.Unlock()
	
	top := atomic.LoadInt64(&d.top)
	b := atomic.LoadInt64(&d.bottom)
	
	if top >= b {
		return Task{}, false // empty
	}
	
	t := d.tasks[top%int64(len(d.tasks))]
	if !atomic.CompareAndSwapInt64(&d.top, top, top+1) {
		return Task{}, false // lost race
	}
	
	return t, true
}

// Len returns approximate size
func (d *WSDeque) Len() int {
	b := atomic.LoadInt64(&d.bottom)
	t := atomic.LoadInt64(&d.top)
	size := b - t
	if size < 0 {
		return 0
	}
	return int(size)
}

// ============================================================================
// ual Work-Stealing (using decoupled views)
// ============================================================================

// WSStack is a work-stealing stack using ual's decoupled views
type WSStack struct {
	stack     *Stack
	ownerView *View // LIFO - owner pops newest
	thiefView *View // FIFO - thief steals oldest
	
	// Pre-allocated buffers for capped stacks (avoids allocation on push)
	buffers   [][]byte
	bufSize   int
	bufHead   int // next buffer for FIFO steal to free
	bufTail   int // next buffer for LIFO push to use
}

// NewWSStack creates a ual-based work-stealing stack (unlimited)
func NewWSStack() *WSStack {
	s := NewStack(LIFO, TypeBytes)
	
	owner := NewView(LIFO)
	owner.Attach(s)
	
	thief := NewView(FIFO)
	thief.Attach(s)
	
	return &WSStack{
		stack:     s,
		ownerView: owner,
		thiefView: thief,
	}
}

// NewWSStackCapped creates a ual-based work-stealing stack with fixed capacity
// Pre-allocates buffers to avoid allocation on push
func NewWSStackCapped(capacity int) *WSStack {
	return NewWSStackCappedWithBufSize(capacity, 64) // default 64 bytes per task
}

// NewWSStackCappedWithBufSize creates a capped stack with custom buffer size
func NewWSStackCappedWithBufSize(capacity, bufSize int) *WSStack {
	s := NewCappedStack(LIFO, TypeBytes, capacity)
	
	owner := NewView(LIFO)
	owner.Attach(s)
	
	thief := NewView(FIFO)
	thief.Attach(s)
	
	// Pre-allocate all buffers
	buffers := make([][]byte, capacity)
	for i := range buffers {
		buffers[i] = make([]byte, bufSize)
	}
	
	return &WSStack{
		stack:     s,
		ownerView: owner,
		thiefView: thief,
		buffers:   buffers,
		bufSize:   bufSize,
		bufHead:   0,
		bufTail:   0,
	}
}

// Push adds a task (owner)
func (ws *WSStack) Push(t Task) bool {
	var data []byte
	
	if ws.buffers != nil {
		// Capped: use pre-allocated buffer at tail
		cap := len(ws.buffers)
		if (ws.bufTail+1)%cap == ws.bufHead && ws.stack.Len() > 0 {
			return false // full
		}
		data = ws.buffers[ws.bufTail%cap]
		encodeTaskInto(t, data)
		ws.bufTail++
	} else {
		// Unlimited: allocate
		data = encodeTask(t)
	}
	
	return ws.stack.Push(data) == nil
}

// Pop removes a task (owner, LIFO)
func (ws *WSStack) Pop() (Task, bool) {
	data, err := ws.ownerView.Pop()
	if err != nil {
		return Task{}, false
	}
	
	// LIFO pop frees tail buffer
	if ws.buffers != nil {
		ws.bufTail--
	}
	
	return decodeTask(data), true
}

// Steal takes a task (thief, FIFO)
func (ws *WSStack) Steal() (Task, bool) {
	data, err := ws.thiefView.Pop()
	if err != nil {
		return Task{}, false
	}
	
	// FIFO steal frees head buffer
	if ws.buffers != nil {
		ws.bufHead++
	}
	
	return decodeTask(data), true
}

// Len returns size
func (ws *WSStack) Len() int {
	return ws.stack.Len()
}

// Simple task encoding (ID as 8 bytes + data)
func encodeTask(t Task) []byte {
	b := make([]byte, 8+len(t.Data))
	encodeTaskInto(t, b)
	return b
}

// encodeTaskInto writes task into existing buffer (no allocation)
func encodeTaskInto(t Task, b []byte) {
	id := t.ID
	for i := 7; i >= 0; i-- {
		b[i] = byte(id & 0xff)
		id >>= 8
	}
	copy(b[8:], t.Data)
}

func decodeTask(b []byte) Task {
	if len(b) < 8 {
		return Task{}
	}
	var id int64
	for i := 0; i < 8; i++ {
		id = (id << 8) | int64(b[i])
	}
	return Task{ID: id, Data: b[8:]}
}

File: rual/src/worksteal.rs
//! Work stealing primitives
//!
//! Two implementations:
//! - `WSDeque`: Traditional Chase-Lev deque (lock-free owner, locked steal)
//! - `WSStack`: ual-native work stealing using decoupled views

use std::sync::atomic::{AtomicI64, AtomicBool, Ordering};
use parking_lot::Mutex;
use crate::{Stack, Perspective, View};
use std::sync::Arc;

/// A unit of work
#[derive(Debug, Clone)]
pub struct Task {
    pub id: i64,
    pub data: Vec<u8>,
}

impl Task {
    pub fn new(id: i64, data: Vec<u8>) -> Self {
        Task { id, data }
    }

    /// Encode to bytes
    pub fn to_bytes(&self) -> Vec<u8> {
        let mut buf = Vec::with_capacity(8 + self.data.len());
        buf.extend_from_slice(&self.id.to_be_bytes());
        buf.extend_from_slice(&self.data);
        buf
    }

    /// Decode from bytes
    pub fn from_bytes(b: &[u8]) -> Option<Self> {
        if b.len() < 8 {
            return None;
        }
        let id = i64::from_be_bytes(b[0..8].try_into().ok()?);
        let data = b[8..].to_vec();
        Some(Task { id, data })
    }
}

// =============================================================================
// Traditional Work-Stealing Deque (Chase-Lev style)
// =============================================================================

/// Chase-Lev work-stealing deque
/// 
/// - Owner pushes and pops from bottom (LIFO)
/// - Thieves steal from top (FIFO)
/// - Lock-free for owner operations, locked steals
pub struct WSDeque {
    tasks: Vec<Mutex<Option<Task>>>,
    bottom: AtomicI64,
    top: AtomicI64,
    capacity: usize,
}

impl WSDeque {
    /// Create a deque with fixed capacity
    pub fn new(capacity: usize) -> Self {
        let mut tasks = Vec::with_capacity(capacity);
        for _ in 0..capacity {
            tasks.push(Mutex::new(None));
        }
        WSDeque {
            tasks,
            bottom: AtomicI64::new(0),
            top: AtomicI64::new(0),
            capacity,
        }
    }

    /// Push a task (owner only)
    pub fn push(&self, task: Task) -> bool {
        let b = self.bottom.load(Ordering::Relaxed);
        let t = self.top.load(Ordering::Acquire);

        if (b - t) as usize >= self.capacity {
            return false; // Full
        }

        let idx = (b as usize) % self.capacity;
        *self.tasks[idx].lock() = Some(task);
        self.bottom.store(b + 1, Ordering::Release);
        true
    }

    /// Pop a task (owner only, LIFO)
    pub fn pop(&self) -> Option<Task> {
        let b = self.bottom.load(Ordering::Relaxed) - 1;
        self.bottom.store(b, Ordering::SeqCst);

        let t = self.top.load(Ordering::SeqCst);

        if t <= b {
            let idx = (b as usize) % self.capacity;
            let task = self.tasks[idx].lock().take();

            if t == b {
                // Last element - race with steal
                if self.top.compare_exchange(
                    t, t + 1,
                    Ordering::SeqCst,
                    Ordering::Relaxed
                ).is_err() {
                    // Lost race to thief
                    self.bottom.store(t + 1, Ordering::Relaxed);
                    return None;
                }
                self.bottom.store(t + 1, Ordering::Relaxed);
            }
            task
        } else {
            // Empty
            self.bottom.store(t, Ordering::Relaxed);
            None
        }
    }

    /// Steal a task (thief, FIFO)
    pub fn steal(&self) -> Option<Task> {
        let t = self.top.load(Ordering::Acquire);
        let b = self.bottom.load(Ordering::Acquire);

        if t >= b {
            return None; // Empty
        }

        let idx = (t as usize) % self.capacity;
        let task = self.tasks[idx].lock().take();

        if self.top.compare_exchange(
            t, t + 1,
            Ordering::SeqCst,
            Ordering::Relaxed
        ).is_err() {
            // Lost race
            // Put it back if we took it
            if task.is_some() {
                *self.tasks[idx].lock() = task;
            }
            return None;
        }

        task
    }

    /// Approximate size
    pub fn len(&self) -> usize {
        let b = self.bottom.load(Ordering::Relaxed);
        let t = self.top.load(Ordering::Relaxed);
        let size = b - t;
        if size < 0 { 0 } else { size as usize }
    }

    /// Check if empty
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

// =============================================================================
// ual Work-Stealing Stack (using decoupled views)
// =============================================================================

/// ual-native work stealing using stack views
///
/// Uses a single stack with two views:
/// - Owner view: LIFO (pops newest)
/// - Thief view: FIFO (steals oldest)
pub struct WSStack {
    stack: Arc<Stack<Vec<u8>>>,
    owner_view: View<Vec<u8>>,
    thief_view: View<Vec<u8>>,
    closed: AtomicBool,
}

impl WSStack {
    /// Create a new work-stealing stack
    pub fn new() -> Self {
        let stack = Arc::new(Stack::new(Perspective::LIFO));
        let owner_view = View::lifo(Arc::clone(&stack));
        let thief_view = View::fifo(Arc::clone(&stack));

        WSStack {
            stack,
            owner_view,
            thief_view,
            closed: AtomicBool::new(false),
        }
    }

    /// Create with capacity
    pub fn with_capacity(capacity: usize) -> Self {
        let stack = Arc::new(Stack::with_capacity(Perspective::LIFO, capacity));
        let owner_view = View::lifo(Arc::clone(&stack));
        let thief_view = View::fifo(Arc::clone(&stack));

        WSStack {
            stack,
            owner_view,
            thief_view,
            closed: AtomicBool::new(false),
        }
    }

    /// Push a task (owner)
    pub fn push(&self, task: Task) -> bool {
        if self.closed.load(Ordering::Relaxed) {
            return false;
        }
        self.owner_view.push(task.to_bytes()).is_ok()
    }

    /// Pop a task (owner, LIFO - gets newest)
    pub fn pop(&self) -> Option<Task> {
        self.owner_view.pop().ok().and_then(|b| Task::from_bytes(&b))
    }

    /// Steal a task (thief, FIFO - gets oldest)
    pub fn steal(&self) -> Option<Task> {
        self.thief_view.pop().ok().and_then(|b| Task::from_bytes(&b))
    }

    /// Get length
    pub fn len(&self) -> usize {
        self.stack.len()
    }

    /// Check if empty
    pub fn is_empty(&self) -> bool {
        self.stack.is_empty()
    }

    /// Close (no more pushes)
    pub fn close(&self) {
        self.closed.store(true, Ordering::Release);
        self.stack.close();
    }

    /// Check if closed
    pub fn is_closed(&self) -> bool {
        self.closed.load(Ordering::Acquire)
    }
}

impl Default for WSStack {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::thread;
    use std::sync::Arc;

    #[test]
    fn test_wsdeque_basic() {
        let deque = WSDeque::new(16);
        
        deque.push(Task::new(1, vec![1]));
        deque.push(Task::new(2, vec![2]));
        deque.push(Task::new(3, vec![3]));

        // Owner pops LIFO (newest first)
        assert_eq!(deque.pop().unwrap().id, 3);
        assert_eq!(deque.pop().unwrap().id, 2);
        assert_eq!(deque.pop().unwrap().id, 1);
        assert!(deque.pop().is_none());
    }

    #[test]
    fn test_wsdeque_steal() {
        let deque = WSDeque::new(16);
        
        deque.push(Task::new(1, vec![]));
        deque.push(Task::new(2, vec![]));
        deque.push(Task::new(3, vec![]));

        // Thief steals FIFO (oldest first)
        assert_eq!(deque.steal().unwrap().id, 1);
        
        // Owner still pops LIFO
        assert_eq!(deque.pop().unwrap().id, 3);
        assert_eq!(deque.pop().unwrap().id, 2);
    }

    #[test]
    fn test_wsstack_basic() {
        let stack = WSStack::new();
        
        stack.push(Task::new(1, vec![1]));
        stack.push(Task::new(2, vec![2]));
        stack.push(Task::new(3, vec![3]));

        // Owner pops LIFO
        assert_eq!(stack.pop().unwrap().id, 3);
        
        // Thief steals FIFO
        assert_eq!(stack.steal().unwrap().id, 1);
        
        // Remaining
        assert_eq!(stack.pop().unwrap().id, 2);
        assert!(stack.pop().is_none());
    }

    #[test]
    fn test_wsstack_concurrent() {
        let stack = Arc::new(WSStack::new());
        
        // Push some initial work
        for i in 0..100 {
            stack.push(Task::new(i, vec![]));
        }

        let stack1 = Arc::clone(&stack);
        let stack2 = Arc::clone(&stack);

        // Owner thread
        let owner = thread::spawn(move || {
            let mut count = 0;
            while let Some(_) = stack1.pop() {
                count += 1;
            }
            count
        });

        // Thief thread
        let thief = thread::spawn(move || {
            let mut count = 0;
            while let Some(_) = stack2.steal() {
                count += 1;
            }
            count
        });

        let owner_count = owner.join().unwrap();
        let thief_count = thief.join().unwrap();

        // Together they should have processed all 100
        assert_eq!(owner_count + thief_count, 100);
    }
}
